Importing library & loading dataset

In [16]:
import pandas as pd
df= pd.read_csv("../raw_data/customer_shopping_behavior.csv")

In [17]:
print(f"Initial shape: {df.shape}")

Initial shape: (10000, 26)


Data Inspection

In [18]:
print(df.head(10))

  Transaction ID Customer ID Purchase Date  Age  Gender       Location  \
0      TXN507808  CUST102534    2023-01-01   51    Male        Chennai   
1      TXN503830  CUST101256    2023-01-01   27  Female         Mumbai   
2      TXN501496  CUST100490    2023-01-01   57    Male      Hyderabad   
3      TXN501538  CUST100500    2023-01-01   22  Female           Pune   
4      TXN504265  CUST101399    2023-01-01   30    Male  Visakhapatnam   
5      TXN509250  CUST102952    2023-01-01   57    Male          Kochi   
6      TXN501814  CUST100596    2023-01-01   42    Male      Hyderabad   
7      TXN500031  CUST100007    2023-01-01   47    Male           Pune   
8      TXN501550  CUST100505    2023-01-01   37  Female      Ahmedabad   
9      TXN505122  CUST101684    2023-01-01   33    Male  Visakhapatnam   

  Online/Offline Online Store     Category Item Purchased  ... Festival/Sale  \
0         Online     Flipkart     Clothing          Kurta  ...           NaN   
1        Offline         

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Transaction ID          10000 non-null  object
 1   Customer ID             10000 non-null  object
 2   Purchase Date           10000 non-null  object
 3   Age                     10000 non-null  int64 
 4   Gender                  10000 non-null  object
 5   Location                10000 non-null  object
 6   Online/Offline          10000 non-null  object
 7   Online Store            7751 non-null   object
 8   Category                10000 non-null  object
 9   Item Purchased          10000 non-null  object
 10  Brand                   10000 non-null  object
 11  Color                   10000 non-null  object
 12  Size                    8965 non-null   object
 13  Quantity                10000 non-null  int64 
 14  Purchase Amount (₹)     10000 non-null  int64 
 15  Dis

In [20]:
df.describe()

,Age,Quantity,Purchase Amount (₹),Discount (%),Shipping Charge (₹),Delivery Time (Days),Review Rating,Previous Purchases
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,38.278100,1.425100,1889.882000,32.916300,41.49210,3.703600,3.874900,3.211400
std,12.133041,0.761346,2102.301836,21.440617,64.30923,2.726223,0.912871,3.951313
min,18.000000,1.000000,67.000000,0.000000,0.00000,0.000000,1.000000,0.000000
25%,28.000000,1.000000,600.000000,12.000000,0.00000,1.000000,3.000000,0.000000
50%,38.000000,1.000000,1207.000000,33.000000,0.00000,4.000000,4.000000,2.000000
75%,49.000000,2.000000,2376.000000,52.000000,79.00000,6.000000,5.000000,4.000000
max,59.000000,4.000000,22306.000000,70.000000,199.00000,8.000000,5.000000,28.000000


In [21]:
df.describe(include='all')

,Transaction ID,Customer ID,Purchase Date,Age,Gender,Location,Online/Offline,Online Store,Category,Item Purchased,...,Festival/Sale,Shipping Charge (₹),Delivery Speed,Delivery Time (Days),Subscription Status,Payment Method,Review Rating,Return Status,Previous Purchases,Frequency of Purchases
count,10000,10000,10000,10000.000000,10000,10000,10000,7751,10000,10000,...,3212,10000.00000,7751,10000.000000,10000,10000,10000.000000,10000,10000.000000,10000
unique,10000,2581,731,NaN,2,20,2,6,3,24,...,7,NaN,3,NaN,2,7,NaN,2,NaN,5
top,TXN507808,CUST101018,2023-08-15,NaN,Male,Pune,Online,Myntra,Clothing,Dress,...,End of Season Sale,NaN,Standard,NaN,No,UPI,NaN,Not Returned,NaN,Weekly
freq,1,29,41,NaN,5449,974,7751,2078,5003,617,...,517,NaN,4945,NaN,6881,4411,NaN,8198,NaN,3487
mean,NaN,NaN,NaN,38.278100,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,41.49210,NaN,3.703600,NaN,NaN,3.874900,NaN,3.211400,NaN
std,NaN,NaN,NaN,12.133041,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,64.30923,NaN,2.726223,NaN,NaN,0.912871,NaN,3.951313,NaN
min,NaN,NaN,NaN,18.000000,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,0.00000,NaN,0.000000,NaN,NaN,1.000000,NaN,0.000000,NaN
25%,NaN,NaN,NaN,28.000000,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,0.00000,NaN,1.000000,NaN,NaN,3.000000,NaN,0.000000,NaN
50%,NaN,NaN,NaN,38.000000,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,0.00000,NaN,4.000000,NaN,NaN,4.000000,NaN,2.000000,NaN
75%,NaN,NaN,NaN,49.000000,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,79.00000,NaN,6.000000,NaN,NaN,5.000000,NaN,4.000000,NaN


In [22]:
df.duplicated().sum()

0

Note:
- Duplicates Not Found. 

# Data Cleaning:
- 1. Null Value Check and Handling.
- 2. Renaming Location column to Cities column.
- 3. Convert Purchase Date to datetime Dtype.

- 1. Null Value Check and Handling.

In [23]:
# Null Value Check
df.isnull().sum().sort_values(ascending=False)

Festival/Sale             6788
Online Store              2249
Delivery Speed            2249
Size                      1035
Transaction ID               0
Purchase Amount (₹)          0
Previous Purchases           0
Return Status                0
Review Rating                0
Payment Method               0
Subscription Status          0
Delivery Time (Days)         0
Shipping Charge (₹)          0
Discount (%)                 0
Quantity                     0
Customer ID                  0
Color                        0
Brand                        0
Item Purchased               0
Category                     0
Online/Offline               0
Location                     0
Gender                       0
Age                          0
Purchase Date                0
Frequency of Purchases       0
dtype: int64

Note:
- Found nulls in:
    - Festival/Sale             6788
    - Online Store              2249
    - Delivery Speed            2249
    - Size                      1035

In [24]:
# Null Value Handling for Festival/Sale

df['Festival/Sale'] = df['Festival/Sale'].fillna('N/A')

In [25]:
# Verifing Null Value Replacement

print(df['Festival/Sale'].isna().sum())

0


In [26]:
# Null Value Handling for Online Store

df.loc[
    (df['Online/Offline'] == 'Offline') & 
    (df['Online Store'].isna()),
    'Online Store'
] = 'Pysical Store'

In [36]:
# Verifing Null Value Replacement

df[['Online/Offline', 'Online Store']].head(10)

,Online/Offline,Online Store
0,Online,Flipkart
1,Offline,Pysical Store
2,Online,Flipkart
3,Online,Myntra
4,Online,Myntra
5,Offline,Pysical Store
6,Online,Myntra
7,Online,Myntra
8,Offline,Pysical Store
9,Online,Amazon


In [28]:
# Null Value Handling for Delivery Speed:

df['Delivery Speed'] = df['Delivery Speed'].fillna('N/A')

In [29]:
# Verifing Null Value Replacement:

print(df['Delivery Speed'].isna().sum())

0


In [30]:
# Checking Unique category column for size:

df.loc[df['Size'].isna(), 'Category'].unique()

array(['Accessories'], dtype=object)

Note:
- All Null values are from the same category.

In [31]:
# Checking the null value category as per the item purchased and size before null handling in size column:

df[df[['Category', 'Item Purchased','Size']].isnull().any(axis=1)][['Category', 'Item Purchased', 'Size']]

,Category,Item Purchased,Size
5,Accessories,Watch,NaN
33,Accessories,Watch,NaN
34,Accessories,Sunglasses,NaN
44,Accessories,Wallet,NaN
47,Accessories,Wallet,NaN
...,...,...,...
9935,Accessories,Belt,NaN
9954,Accessories,Watch,NaN
9974,Accessories,Watch,NaN
9988,Accessories,Sunglasses,NaN


Note:
- No specific sub-category relation found for the nulls.
- Thus, filling up with N/A

In [ ]:
# Null value handling for size column:

df['Size'] = df['Size'].fillna('N/A')

In [33]:
# Verifying Null Value Replacement:

print(df['Size'].isna().sum())

0


- 2. Renaming the Location column with Cities column.

In [ ]:
df.rename(columns={'Location': 'Cities'}, inplace=True)

- 3. Convert Purchase Date to datetime Dtype

In [34]:
df['Purchase Date'] = pd.to_datetime(df['Purchase Date'], errors = 'coerce')

print(df['Purchase Date'].dtype)

datetime64[ns]


Feature Engineering:
- 1. Creating a New Region Column.
- 2. Extracting Parts of date.
- 3. Creating Age group 

- 1. Creating a New Region Column:

In [49]:
# All unique city names:

df['Cities'].unique()

array(['Chennai', 'Mumbai', 'Hyderabad', 'Pune', 'Visakhapatnam', 'Kochi',
       'Ahmedabad', 'Chandigarh', 'Delhi', 'Bangalore', 'Indore',
       'Jaipur', 'Kolkata', 'Patna', 'Lucknow', 'Nagpur', 'Noida',
       'Surat', 'Bhubaneswar', 'Gurgaon'], dtype=object)

In [ ]:
# Mapping the cities

region_map = {
    'Delhi': 'North',
    'Noida': 'North',
    'Gurgaon': 'North',
    'Chandigarh': 'North',
    'Jaipur': 'North',
    'Lucknow': 'North',

    'Chennai': 'South',
    'Hyderabad': 'South',
    'Visakhapatnam': 'South',
    'Kochi': 'South',
    'Bangalore': 'South',

    'Mumbai': 'West',
    'Pune': 'West',
    'Ahmedabad': 'West',
    'Nagpur': 'West',
    'Surat': 'West',

    'Kolkata': 'East',
    'Patna': 'East',
    'Bhubaneswar': 'East',

    'Indore': 'Central'
}

df['Region'] = df['Cities'].map(region_map)

In [ ]:
# Popping out the region column:

region = df.pop('Region')

# Finding the position of the Cities column:

loc = df.columns.get_loc('Cities')

# Inserting Region column immediately after Cities column:

df.insert(loc + 1, 'Region', region)


- 2. Extracting Parts of date and inserting them immediately after the Purchase Date column:

In [62]:
# Finding the position of the Purchase Date column:
loc = df.columns.get_loc('Purchase Date')


# Extracting Year and inserting it immediately after the Purchase Date column:
df.insert(loc + 1, 'Year', df['Purchase Date'].dt.year)


# Extracting Month and inserting it immediately after the Year column:
df.insert(loc + 2, 'Month', df['Purchase Date'].dt.month)


# Extracting Day and inserting it immediately after the Month column:
df.insert(loc + 3, 'Day', df['Purchase Date'].dt.day)


# Extracting Weekday Name and inserting it immediately after the Day column:
df.insert(loc + 4, 'Weekday', df['Purchase Date'].dt.day_name())


- 3. Creating Age group:

In [65]:
def age_group(age):
    if age < 18:
        return 'Under 18'
    elif age <= 25:
        return '18-25'
    elif age <= 35:
        return '26-35'
    elif age <= 45:
        return '36-45'
    elif age <= 60:
        return '46-60'
    else:
        return 'Above 60'
    
# Finding the position of the age column:
loc = df.columns.get_loc('Age')

# Inserting the Age Group column immediately after the age column:
df.insert(loc + 1, 'Age Group', df['Age'].apply(age_group))

In [66]:
df.head()

,Transaction ID,Customer ID,Purchase Date,Year,Month,Day,Weekday,Age,Age Group,Gender,...,Festival/Sale,Shipping Charge (₹),Delivery Speed,Delivery Time (Days),Subscription Status,Payment Method,Review Rating,Return Status,Previous Purchases,Frequency of Purchases
0,TXN507808,CUST102534,2023-01-01,2023,1,1,Sunday,51,46-60,Male,...,N/A,0,Same Day,1,Yes,Cash on Delivery,4,Not Returned,0,Fortnightly
1,TXN503830,CUST101256,2023-01-01,2023,1,1,Sunday,27,26-35,Female,...,N/A,0,N/A,0,No,UPI,5,Not Returned,0,Fortnightly
2,TXN501496,CUST100490,2023-01-01,2023,1,1,Sunday,57,46-60,Male,...,N/A,0,Standard,7,Yes,UPI,3,Not Returned,0,Rarely
3,TXN501538,CUST100500,2023-01-01,2023,1,1,Sunday,22,18-25,Female,...,N/A,0,Standard,6,Yes,Net Banking,4,Not Returned,0,Fortnightly
4,TXN504265,CUST101399,2023-01-01,2023,1,1,Sunday,30,26-35,Male,...,N/A,99,Express,2,No,Debit Card,4,Not Returned,0,Weekly
